packages/nouns-contracts/contracts/NounsAuctionHouseV2.sol

```diff
@@ -0,0 +1,511 @@
+            _safeTransferETHWithFallback(owner(), _auction.amount);
+        }
+
+        settlementHistory[_auction.nounId] = SettlementState({
+            blockTimestamp: uint32(block.timestamp),
+            amount: ethPriceToUint64(_auction.amount),
+            winner: _auction.bidder
+        });
+
+        emit AuctionSettled(_auction.nounId, _auction.bidder, _auction.amount);
+    }
+
+    /**
+     * @notice Transfer ETH. If the ETH transfer fails, wrap the ETH and try send it as WETH.
+     */
+    function _safeTransferETHWithFallback(address to, uint256 amount) internal {
+        if (!_safeTransferETH(to, amount)) {
+            IWETH(weth).deposit{ value: amount }();
+            IERC20(weth).transfer(to, amount);
+        }
+    }
+
+    /**
+     * @notice Transfer ETH and return the success status.
+     * @dev This function only forwards 30,000 gas to the callee.
+     */
+    function _safeTransferETH(address to, uint256 value) internal returns (bool) {
+        bool success;
+        assembly {
+            success := call(30000, to, value, 0, 0, 0, 0)
+        }
+        return success;
+    }
+
+    /**
+     * @notice Set historic prices; only callable by the owner, which in Nouns is the treasury (timelock) contract.
+     * @dev This function lowers auction price accuracy from 18 decimals to 10 decimals, as part of the price history
+     * bit packing, to save gas.
+     * @param settlements The list of historic prices to set.
+     */
+    function setPrices(Settlement[] memory settlements) external onlyOwner {
+        uint256[] memory nounIds = new uint256[](settlements.length);
+        uint256[] memory prices_ = new uint256[](settlements.length);
+
+        for (uint256 i = 0; i < settlements.length; ++i) {
+            settlementHistory[settlements[i].nounId] = SettlementState({
+                blockTimestamp: settlements[i].blockTimestamp,
+                amount: ethPriceToUint64(settlements[i].amount),
+                winner: settlements[i].winner
+            });
+
+            nounIds[i] = settlements[i].nounId;
+            prices_[i] = settlements[i].amount;
+        }
+
+        emit HistoricPricesSet(nounIds, prices_);
+    }
+
+    /**
+     * @notice Warm up the settlement state for a list of Noun IDs.
+     * @dev Helps lower the gas cost of auction settlement when storing settlement data
+     * thanks to the state slot being non-zero.
+     * @dev Only writes to slots where blockTimestamp is zero, meaning it will not overwrite existing data.
+     * @param nounIds The list of Noun IDs whose settlement slot to warm up.
+     */
+    function warmUpSettlementState(uint256[] calldata nounIds) external {
+        for (uint256 i = 0; i < nounIds.length; ++i) {
+            if (settlementHistory[nounIds[i]].blockTimestamp == 0) {
+                settlementHistory[nounIds[i]] = SettlementState({ blockTimestamp: 1, amount: 0, winner: address(0) });
+            }
+        }
+    }
```

Comment:
It's likely worth optimizing this function since it may be called to warm up thousands of slots.

e.g.
```solidity
function warmUpSettlementState(uint256[] calldata nounIds) external {
    uint256 nounIdCount = nounIds.length;
    SettlementState storage settlementState;
    for (uint256 i; i < nounIdCount;) {
        settlementState = settlementHistory[nounIds[i]];
        if (settlementState.blockTimestamp == 0) {
            settlementState.blockTimestamp = 1;
        }
        unchecked {
            ++i;
        }
    }
}
```